sub-packages/bionemo-dnadl/src/bionemo/dnadl/io/dataset.py

```diff
@@ -0,0 +1,189 @@
+    ):
+        """Instantiate the class.
+
+        Args:
+            genome: Genome object
+            bed_file_df: A bed file in the form of a polars DataFrame with chrom, start, and end columns at least
+            context_length: Size of the window
+            chr_bed_to_fasta_map: Mapping between chromosome names in the bed file and the fasta file
+            **kwargs: Additional arguments
+        """
+        super().__init__(**kwargs)
+        assert {"chrom", "start", "end"}.issubset(
+            bed_file_df.columns
+        ), "bed file df must contain columns ['chrom', 'start', 'end']"
+        self.bed_file_df = bed_file_df
+
+        # if the chromosome name in the bed file is different than the keyname in the fasta
+        # can remap on the fly
+        self.chr_bed_to_fasta_map = chr_bed_to_fasta_map
+        self.context_length = context_length
+        self.genome = genome
+
+    def __len__(self) -> int:
+        """Return the length of the dataset."""
+        return len(self.bed_file_df)
+
+    def get_interval(self, ind: int) -> GenomeInterval:
+        """Get the interval from the bed file."""
+        chrom = self.chr_bed_to_fasta_map.get(self.bed_file_df["chrom"][ind], self.bed_file_df["chrom"][ind])
+        genome_interval = GenomeInterval(chrom, self.bed_file_df["start"][ind], self.bed_file_df["end"][ind])
+        if self.context_length is not None:
+            return self.genome.adjust_interval(genome_interval, self.context_length)
+        return genome_interval
+
+    def get_dna_sequence(self, ind: int) -> str:
+        """Get the DNA sequence from the genome."""
+        genome_interval = self.get_interval(ind)
+        return self.genome.extract_sequence(genome_interval)
+
+
+class VCFDataset(GenomeDataset):
+    """Dataset that creates sequences based on a VCF. It needs a genome and a bed file to specify the windows."""
+
+    def __init__(
+        self,
+        vcf_file: str | Path,
+        genome_interval_dataset: GenomeIntervalDataset,
+        sample_ids: Optional[List[SampleId]] = None,
+        **kwargs: Any,
+    ):
+        """Initialize the class."""
+        super().__init__(**kwargs)
+        self.genome_interval_dataset = genome_interval_dataset
+        self.vcf_file = Path(vcf_file)
+        assert self.vcf_file.exists(), "path to VCF file must exist"
+
+        if sample_ids is None:
+            self._populate_sample_ids_from_vcf()
+        else:
+            self.sample_ids = sample_ids
+
+    def _populate_sample_ids_from_vcf(self) -> None:
+        with pysam.VariantFile(str(self.vcf_file)) as vcf_file:
+            self.sample_ids = list(vcf_file.header.samples)
+
+    def __len__(self) -> int:
+        """Return the length of the dataset."""
+        return len(self.genome_interval_dataset) * len(self.sample_ids)
+
+    def get_dna_sequence(self, ind: int) -> str:
+        """Get an item from the dataset."""
+        interval_index = ind // len(self.sample_ids)
```

Comment:
this is a great place to surface and document all the edge-cases you handle when mutating the reference sequence.